infra/Pos/DHT/Real/Real.hs

```diff
@@ -138,40 +140,40 @@ getKnownPeersImpl = do
     (inst, initialPeers, explicitInitial) <-
         KademliaDHT $
         (,,) <$> ask <*> asks kdiInitialPeers <*> asks kdiExplicitInitial
-    peers <- liftIO $ K.dumpPeers $ kdiHandle inst
+    buckets <- liftIO (K.viewBuckets $ kdiHandle inst)
     let initPeers = bool [] initialPeers explicitInitial
-    filter ((/= myId) . dhtNodeId) <$> extendPeers inst myId initPeers peers
+    filter ((/= myId) . dhtNodeId) <$> extendPeers inst myId initPeers buckets
   where
-    extendPeers inst myId initial peers =
+    extendPeers
+        :: MonadIO m1
+        => KademliaDHTInstance
+        -> DHTKey
+        -> [DHTNode]
+        -> [[(K.Node DHTKey, Int64)]]
+        -> m1 [DHTNode]
+    extendPeers inst myId initial buckets =
         map snd .
         HM.toList .
         HM.delete myId .
         flip (foldr $ \n -> HM.insert (dhtNodeId n) n) initial .
         HM.fromList . map (\(toDHTNode -> n) -> (dhtNodeId n, n)) <$>
-        (updateCache inst =<< selectSufficientNodes inst myId peers)
-    selectSufficientNodes inst myId l =
-        if enhancedMessageBroadcast /= (0 :: Int)
-            then concatMapM (getPeersFromBucket enhancedMessageBroadcast inst)
-                     (splitToBuckets (kdiHandle inst) myId l)
-            else return l
-    bucketIndex origin x =
-        length . takeWhile (== False) <$> K.distance origin (K.nodeId x)
-    insertId origin i hm = do
-        bucket <- bucketIndex origin i
-        return $ HM.insertWith (++) bucket [i] hm
-    splitToBuckets kInst origin peers =
-        flip K.usingKademliaInstance kInst $
-        HM.elems <$> foldrM (insertId origin) HM.empty peers
-    getPeersFromBucket p inst bucket = do
-        cache <- atomically $ readTVar $ kdiKnownPeersCache inst
-        let fromCache = tryTake p $ intersect cache bucket
-        return $ fromCache ++ tryTake (p - length fromCache) (bucket \\ cache)
-    tryTake x l
-        | length l < x = l
-        | otherwise = take x l
-    updateCache inst peers = do
-        atomically $ writeTVar (kdiKnownPeersCache inst) peers
-        return peers
+        (updateCache inst $ concatMap getPeersFromBucket buckets)
+
+    getPeersFromBucket
+        :: [(K.Node DHTKey, Int64)]
+        -> [K.Node DHTKey]
+    getPeersFromBucket bucket
+        | null bucket = []
+        | otherwise =
+            let peers = filter ((< enhancedMessageTimeout) . snd) bucket in
+            if null peers then
+                [fst $ minimumBy (\x y -> compare (snd x) (snd y)) bucket]
```

Comment:
Although Artyom's example is better, `minimumBy (compare `on` snd)` also works. `comparing` exists for this, but the `on` combinator is useful to know.